src/libbson/tests/test-bson-vector.c

```diff
@@ -0,0 +1,1400 @@
+      {
+         bson_array_builder_t *array_builder = bson_array_builder_new ();
+         if (!bson_array_builder_append_vector_elements (array_builder, &iter)) {
+            test_error ("test '%s' should be valid but failed array-from-vector conversion",
+                        test_case->test_description);
+         }
+         BSON_ASSERT (bson_array_builder_build (array_builder, &array_from_vector));
+         bson_array_builder_destroy (array_builder);
+      }
+
+      // (TODO for DRIVERS-3095, DRIVERS-3097) Due to loosely defined element types and rounding behavior in the test
+      // vectors, this comparison can't be exact. Temporary special cases are needed for float32 and packed_bits.
+
+      if (BSON_ITER_HOLDS_VECTOR_FLOAT32 (&iter)) {
+         // float32 special case: We need to handle the nonstandard "inf" and "-inf" forms, and
+         // due to underspecified rounding and conversion rules we compare value inexactly.
+
+         bson_iter_t actual_iter, expected_iter;
+         BSON_ASSERT (bson_iter_init (&actual_iter, &array_from_vector));
+         BSON_ASSERT (bson_iter_init (&expected_iter, test_case->test_vector_array));
+
+         for (size_t i = 0;; i++) {
+            bool actual_next = bson_iter_next (&actual_iter);
+            bool expected_next = bson_iter_next (&expected_iter);
+            if (!actual_next && !expected_next) {
+               break;
+            } else if (!actual_next) {
+               test_error ("converted array is shorter than expected");
+            } else if (!expected_next) {
+               test_error ("converted array is longer than expected");
+            }
+
+            if (!BSON_ITER_HOLDS_DOUBLE (&actual_iter)) {
+               test_error ("converted array element %d has unexpected type, should be double", (int) i);
+            }
+            double actual_double = bson_iter_double (&actual_iter);
+
+            double expected_double;
+            if (BSON_ITER_HOLDS_UTF8 (&expected_iter) && 0 == strcmp ("inf", bson_iter_utf8 (&expected_iter, NULL))) {
+               expected_double = INFINITY;
+            } else if (BSON_ITER_HOLDS_UTF8 (&expected_iter) &&
+                       0 == strcmp ("-inf", bson_iter_utf8 (&expected_iter, NULL))) {
+               expected_double = -INFINITY;
+            } else if (BSON_ITER_HOLDS_DOUBLE (&expected_iter)) {
+               expected_double = bson_iter_double (&expected_iter);
+            } else {
+               test_error ("test-vector array element %d has unexpected type, should be double, 'inf', or '-inf'.",
+                           (int) i);
+            }
+
+            bool is_sorta_equal = false;
+            if (expected_double != expected_double) {
+               // Expect NaN, any type is fine.
+               if (actual_double != actual_double) {
+                  is_sorta_equal = true;
+               }
+            } else if (expected_double == 0 || expected_double * 0 != 0) {
```

Comment:
```suggestion
            } else if (expected_double == 0.0 || expected_double * 0.0 != 0.0) {
```
Addresses `-Wdouble-promotion` warnings and better indicates these are floating-point equality comparisons (type consistency).